File: frontend/lib/schema/notification.ts
import { z } from "zod";

// ──────────────────────────────────────────────
// Enumerações
// ──────────────────────────────────────────────

export const NotificationTypeEnum = z.enum(["failure", "completion"]);
export type NotificationType = z.infer<typeof NotificationTypeEnum>;

// ──────────────────────────────────────────────
// Respostas
// ──────────────────────────────────────────────

export const NotificationResponseSchema = z.object({
  id: z.string().uuid(),
  user_id: z.string(),
  execution_id: z.string().nullable().default(null),
  type: NotificationTypeEnum,
  title: z.string().max(255),
  message: z.string(),
  read: z.boolean().default(false),
  created_at: z.string().datetime()
});
export type NotificationResponse = z.infer<typeof NotificationResponseSchema>;

export const NotificationListResponseSchema = z.object({
  notifications: z.array(NotificationResponseSchema).default([]),
  unread_count: z.number().min(0).default(0),
  total_count: z.number().min(0).default(0)
});
export type NotificationListResponse = z.infer<typeof NotificationListResponseSchema>;

// ──────────────────────────────────────────────
// Ações
// ──────────────────────────────────────────────

export const MarkNotificationReadRequestSchema = z.object({
  notification_ids: z.array(z.string()).min(1)
});
export type MarkNotificationReadRequest = z.infer<typeof MarkNotificationReadRequestSchema>;

export const MarkNotificationReadResponseSchema = z.object({
  marked_count: z.number().min(0),
  unread_remaining: z.number().min(0)
});
export type MarkNotificationReadResponse = z.infer<typeof MarkNotificationReadResponseSchema>;

// ──────────────────────────────────────────────
// Criação interna (não exposta na API pública)
// ──────────────────────────────────────────────

export const CreateNotificationSchema = z.object({
  user_id: z.string(),
  execution_id: z.string().nullable().default(null),
  type: NotificationTypeEnum,
  title: z.string().max(255),
  message: z.string()
});
export type CreateNotification = z.infer<typeof CreateNotificationSchema>;

File: _legacy-v1-python/app/models/notification.py
"""
AdCraft — Notification Models
==============================
Pydantic schemas para request/response de notificações.

Referência: PRD v1.0 — Seção 7 (tabela notifications)

Notificações são criadas internamente pelo backend e publicadas
via Supabase Realtime. Apenas dois tipos:
  - failure:    erro em algum nó da execução
  - completion: fluxo concluído com sucesso
"""

from __future__ import annotations

from datetime import datetime
from enum import Enum
from typing import Optional

from pydantic import BaseModel, Field


# ──────────────────────────────────────────────
# Enumerações
# ──────────────────────────────────────────────

class NotificationType(str, Enum):
    """Tipos de notificação suportados."""
    FAILURE = "failure"
    COMPLETION = "completion"


# ──────────────────────────────────────────────
# Respostas
# ──────────────────────────────────────────────

class NotificationResponse(BaseModel):
    """Representação de notificação retornada pela API."""

    id: str = Field(..., description="UUID da notificação")
    user_id: str
    execution_id: Optional[str] = Field(
        default=None, description="UUID da execução associada"
    )
    type: NotificationType = Field(
        ..., description="Tipo: failure ou completion"
    )
    title: str = Field(
        ..., max_length=255, description="Título curto da notificação"
    )
    message: str = Field(..., description="Corpo completo da mensagem")
    read: bool = Field(default=False, description="Se já foi lida")
    created_at: datetime


class NotificationListResponse(BaseModel):
    """Lista de notificações com contagem de não lidas."""

    notifications: list[NotificationResponse] = Field(
        default_factory=list
    )
    unread_count: int = Field(
        default=0, ge=0, description="Total de notificações não lidas"
    )
    total_count: int = Field(
        default=0, ge=0, description="Total de notificações"
    )


# ──────────────────────────────────────────────
# Ações
# ──────────────────────────────────────────────

class MarkNotificationReadRequest(BaseModel):
    """Payload para marcar notificação(ões) como lida(s)."""

    notification_ids: list[str] = Field(
        ..., min_length=1,
        description="UUIDs das notificações a marcar como lidas"
    )


class MarkNotificationReadResponse(BaseModel):
    """Resposta de marcação de notificações como lidas."""

    marked_count: int = Field(
        ..., ge=0, description="Quantidade marcada como lida"
    )
    unread_remaining: int = Field(
        ..., ge=0, description="Quantidade restante não lida"
    )


# ──────────────────────────────────────────────
# Criação interna (não exposta na API pública)
# ──────────────────────────────────────────────

class CreateNotification(BaseModel):
    """
    Schema interno para criação de notificação pelo backend.

    NÃO é exposto como endpoint da API — usado pelo sistema
    de orquestração ao detectar conclusão ou falha.
    """

    user_id: str = Field(..., description="UUID do usuário a notificar")
    execution_id: Optional[str] = Field(
        default=None, description="UUID da execução relacionada"
    )
    type: NotificationType = Field(..., description="failure ou completion")
    title: str = Field(
        ..., max_length=255,
        description="Título da notificação"
    )
    message: str = Field(
        ..., description="Corpo da mensagem com detalhes do evento"
    )
